### PR TITLE
[stable/jenkins] helm chart best practices

### LINKED ISSUE #13132: [stable/jenkins] chart naming convention best practices
**Is this a request for help?**:
YES!
---

**Which chart**:
stable/jenkins

This is a call for contributors!

As chart maintainers, we have been working towards making the stable/jenkins chart follow chart best practices, while attempting to simplify it.
Over time, some breaking changes have been introduced to the chart. As this chart is in major version 0, the major version was not bumped, which is confusing as the chart is in the stable repository.
This has led to some discussions around these changes, and we have decided to move towards a 1.0 release.

We would like to know if someone is willing to pick up this task and make some improvements to the Jenkins chart.
The goal is to make the Jenkins chart follow [naming conventions best practices](https://helm.sh/docs/chart_best_practices/#naming-conventions).

Thanks!

### PR DESCRIPTION
**Pull request by torstenwalter**
#### What this PR does / why we need it:

Rename values to match [helm chart naming conventions](https://helm.sh/docs/chart_best_practices/#naming-conventions)

- [x] renamed all `Master.*` keys
- [x] renamed all `Agent.*` keys
- [x] renamed all `Persistence.*` keys 
- [x] renamed all `NetworkPolicy.*` keys 

Implement [RBAC best practices](https://helm.sh/docs/chart_best_practices/#role-based-access-control)

- [x] service account is generated by default, if set to false then a name for the service account can be specified. (if not set it defaults to `default`
- [x] `rbac.create` is set to true and necessary rbac resources are generated.

Implement recommended [standard labels](https://helm.sh/docs/chart_best_practices/#standard-labels):

- [x] `app.kubernetes.io/name`
- [x] `helm.sh/chart`
- [x] `app.kubernetes.io/managed-by`
- [x] `app.kubernetes.io/instance`

Others:

- [x] removed support for `masterCpu` / `masterMemory` this can be configured via `resources`
- [x] simplify code for `securityRealm` and `authorizationStrategy` by specifying defaults within `values.yaml`

#### Which issue this PR fixes
  - fixes #13132

#### Special notes for your reviewer:
Renaming all those values is tedious and it's easy to miss one especially since not all values are defined in `values.yaml`

This is a very large PR as it changes nearly every value and implements best practices for a 1.0.0 release. Please let me know if other changes are required.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/jenkins/values.yaml [hunk 1]
```diff
@@ -56,125 +58,126 @@ Master:
...
+  # slaveListenerServiceType is the Kubernetes Service type for the JNLP slave service,
   # either 'LoadBalancer', 'NodePort', or 'ClusterIP'
   # Note if you set this to 'LoadBalancer', you *must* define annotations to secure it. By default
   # this will be an external load balancer and allowing inbound 0.0.0.0/0, a HUGE
   # security risk:  https://github.com/kubernetes/charts/issues/1341
-  SlaveListenerServiceType: ClusterIP
-  SlaveListenerServiceAnnotations: {}
+  slaveListenerServiceType: "ClusterIP"
+  slaveListenerServiceAnnotations: {}
+  slaveKubernetesNamespace:
```
Comment (maorfr) [inline stable/jenkins/values.yaml]: i like it that this was also added in this PR! :)

**Comment by maorfr** [comment]
@torstenwalter, this looks amazing!!!!!!!!!!!!
Thank you so much for picking this up and delivering within the same day!

I'll also let @viglesiasce take a look at this :)

**Comment by viglesiasce** [comment]
This looks great!

What do you think about adding a check for some of the old values that spits out an error instead of installing the chart? I'd like to prevent people from upgrading to this version and causing issues in their setup.

**Comment by torstenwalter** [comment]
@viglesiasce Adding a check shouldn't be a problem. How do you like to cause an error?

**Comment by torstenwalter** [comment]
@viglesiasce @maorfr: I am now checking for old keys and fail in case they are present.

To get better a better feeling for this PR and also future ones a `ci/` folder with values files has been added. 

> Charts may have multiple custom values files matching the glob pattern '*-values.yaml' in a directory named 'ci' in the root of the chart's directory. The chart is linted for each of these files. If no custom values file is present, the chart is linted with defaults.
https://github.com/helm/chart-testing/blob/acfb89768ede45c25fa6fa56ae345d23f68e4f82/doc/ct_lint.md

File: stable/jenkins/templates/deprecation.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+   {{ fail "Agent.* values have been renamed, please check the documentation" }}
+{{- end }}
+
+{{- if .Values.Persistence }}
+   {{ fail "Persistence.* values have been renamed, please check the documentation" }}
+{{- end }}
+
+{{- if .Values.NetworkPolicy }}
+   {{ fail "NetworkPolicy.* values have been renamed, please check the documentation" }}
+{{- end }}
```
Comment (maorfr) [inline stable/jenkins/templates/deprecation.yaml]: Missing new line?

**Comment by maorfr** [comment]
Nicely done @torstenwalter!!

File: stable/jenkins/templates/deprecation.yaml (hunk 2 above)
Comment (maorfr) [reply stable/jenkins/templates/deprecation.yaml]: And maybe we should add something similar for the rbac changes?

Comment (torstenwalter) [reply stable/jenkins/templates/deprecation.yaml]: added checks for all values described in `README.md` which have been renamed. That should avoid a lot of unnecessary questions / issues.

**Comment by maorfr** [comment]
hey @torstenwalter,

are we good to merge?
we will need a rebase here.

**Comment by torstenwalter** [comment]
@maorfr From my side it's okay. Should I rebase the branch on latest master?

**Comment by maorfr** [comment]
yes please

**Comment by torstenwalter** [comment]
@maorfr rebase done

**Comment by maorfr** [comment]
/lgtm

this is great! @torstenwalter, welcome to the team :)

**Comment by torstenwalter** [comment]
When I lint the chart locally with [ct](https://github.com/helm/chart-testing) v2.3.0 the `ci/*-values.yaml` values the output is as follows:

```
Linting chart with values file 'stable/jenkins/ci/casc-values.yaml'...

==> Linting stable/jenkins
Lint OK

1 chart(s) linted, no failures

Linting chart with values file 'stable/jenkins/ci/default-values.yaml'...

==> Linting stable/jenkins
Lint OK
```

As `gcr.io/kubernetes-charts-ci/test-image:v3.2.0` contains  v2.2.0 of `ct` the output is different, but it looks like that both values files are linted:

```
I0422 11:38:34.137] Validating maintainers...
I0422 11:38:42.211] ==> Linting stable/jenkins
I0422 11:38:42.211] Lint OK
I0422 11:38:42.211]
I0422 11:38:42.212] 1 chart(s) linted, no failures
I0422 11:38:42.297] ==> Linting stable/jenkins
I0422 11:38:42.297] Lint OK
```

I think this could help to get feedback for different values, especially if a value is used to decide if a yaml file is rendered or not.

**Pull request merged**